smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/metrics/MetricDecorator.java

```diff
@@ -0,0 +1,39 @@
+package io.smallrye.reactive.messaging.metrics;
+
+import java.util.function.Consumer;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.Tag;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+
+import io.smallrye.reactive.messaging.PublisherDecorator;
+
+@ApplicationScoped
+public class MetricDecorator implements PublisherDecorator {
+
+    @Inject
+    private Instance<MetricRegistry> registryInstance;
+
+    @Override
+    public PublisherBuilder<? extends Message> decoratePublisher(PublisherBuilder<? extends Message> publisher,
+            String channelName) {
+        if (registryInstance.isResolvable()) {
+            return publisher.peek(incrementCount(channelName));
+        } else {
+            return publisher;
+        }
+    }
+
+    private <T extends Message> Consumer<T> incrementCount(String channelName) {
+        MetricRegistry registry = registryInstance.get();
```

Comment:
can we save the retrieval and cache the registry.
Same question for the counter?

Should it be a `Consumer<? extends Message<P>>` ?

Reply:
Not actually sure. As it's a private method and we're passing the return value to `peek()` which is expecting `Consumer<? super Message>`, it could be `Consumer<Message>` or maybe `Consumer<Object>` would make more sense as it actually doesn't care at all about what the object it consumes is.

Reply:
As a side note, it looks like you use the raw `Message` type everywhere, so I've continued that.

Reply:
ok.